src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreDataSource.cs

```diff
@@ -84,10 +101,18 @@ internal class PackageRestoreDataSource : ChainedProjectValueDataSourceBase<Rest
             _fileSystem = fileSystem;
             _logger = logger;
             _projectDependentFileChangeNotificationService = projectDependentFileChangeNotificationService;
+            _solutionRestoreService4 = solutionRestoreService4;
         }
 
         protected override IDisposable? LinkExternalInput(ITargetBlock<IProjectVersionedValue<RestoreData>> targetBlock)
         {
+            // Register before this project receives any data flows containing possible nominations.
+            // This is needed because we need to register before any nuget restore or before the solution load.
+            _project.Services.FaultHandler.Forget(
```

Comment:
I wonder we might register us too later in the initialization time.  Basically, if the restore data source is not established earlier enough, the project won't be registered into NuGet, therefor it won't block project registered earlier to be restored. @davkean : you might recommend a better place to do registration, or you think this concern doesn't stay due to other conditions.

Reply:
we don't need resolve this before merging this PR. it is for future discussion only.